packages/squiggle-lang/__tests__/SqProject/SqProject_imports_test.ts

```diff
@@ -1,81 +1,221 @@
+        "./lib": "export x = 5",
+      }),
+    });
     project.setSource(
       "main",
       `
 import './lib' as lib
 lib.x`
     );
 
-    expect(
-      project.runWithImports("main", async (id) => {
-        throw new Error(`Unknown id ${id}`);
-      })
-    ).rejects.toThrow();
-
-    expect(
-      project.runWithImports("main", async (id) => {
-        if (id === "./lib") {
-          return "x = 5";
-        }
-        throw new Error(`Unknown id ${id}`);
-      })
-    ).resolves.toBe(undefined);
+    await project.run("main");
+
+    expect(project.getResult("main").ok).toEqual(true);
+    expect(project.getResult("main").value.toString()).toEqual("5");
+  });
+
+  describe("Another import test", () => {
+    const project = SqProject.create({
+      linker: buildNaiveLinker(),
+    });
+
+    project.setSource(
+      "first",
+      `
+      import 'common' as common
+      x = 1
+    `
+    );
+    expect(project.getDependencies("first")).toEqual(["common"]);
+
+    project.setSource("common", "export common = 0");
+    project.setSource(
+      "second",
+      `
+      import 'common' as common
+      y = 2
+    `
+    );
+    project.setContinues("second", ["first"]);
+    expect(project.getDependencies("second")).toEqual(["first", "common"]);
+
+    project.setSource("main", "z=3; y");
+    project.setContinues("main", ["second"]);
+
+    test("test result", async () => {
+      expect(await runFetchResult(project, "main")).toBe("Ok(2)");
+    });
+    test("test bindings", async () => {
+      // bindings from continues are not exposed!
+      expect(await runFetchBindings(project, "main")).toBe("{z: 3}");
+    });
+  });
+
+  test("Cyclic imports", async () => {
+    const project = SqProject.create({
+      linker: buildNaiveLinker({
+        bar: `
+          import "foo" as foo
+          y = 6
+        `,
+        foo: `
+          import "bar" as bar
+          x = 5
+        `,
+      }),
+    });
+    project.setSource(
+      "main",
+      `
+        import "foo" as foo
+        foo.x
+      `
+    );
+
+    await project.run("main");
+
+    expect(project.getResult("main").ok).toEqual(false);
+    expect(project.getResult("main").value.toString()).toEqual(
+      "Cyclic import foo"
+    );
+  });
+
+  test("Self-import", async () => {
+    const project = SqProject.create({
+      linker: buildNaiveLinker(),
+    });
+    project.setSource(
+      "main",
+      `
+        import "main" as self
+        self
+      `
+    );
+
+    await project.run("main");
+
+    expect(project.getResult("main").ok).toEqual(false);
+    expect(project.getResult("main").value.toString()).toEqual(
+      "Cyclic import main"
+    );
+  });
+
+  test("Diamond shape", async () => {
+    const project = SqProject.create({
+      linker: buildNaiveLinker({
+        common: `
+          export x = 10
+        `,
+        bar: `
+          import "common" as common
+          export x = common.x * 2
+        `,
+        foo: `
```

Comment:
minor - but it could be nice to name these based on the parts of the diamond they represent